src/dataflow/src/server/compute_state.rs

```diff
@@ -0,0 +1,565 @@
+use mz_dataflow_types::logging::LoggingConfig;
+use mz_dataflow_types::{DataflowError, PeekResponse, TailResponse};
+use mz_expr::GlobalId;
+use mz_repr::Timestamp;
+
+use crate::activator::RcActivator;
+use crate::arrangement::manager::{TraceBundle, TraceManager};
+use crate::logging;
+use crate::logging::materialized::MaterializedEvent;
+use crate::operator::CollectionExt;
+use crate::server::PendingPeek;
+use crate::sink::SinkBaseMetrics;
+
+/// Worker-local state that is maintained across dataflows.
+///
+/// This state is restricted to the COMPUTE state, the deterministic, idempotent work
+/// done between data ingress and egress.
+pub struct ComputeState {
+    /// The traces available for sharing across dataflows.
+    pub traces: TraceManager,
+    /// Tokens that should be dropped when a dataflow is dropped to clean up
+    /// associated state.
+    pub dataflow_tokens: HashMap<GlobalId, Box<dyn Any>>,
+    /// Shared buffer with TAIL operator instances by which they can respond.
+    ///
+    /// The entries are pairs of sink identifier (to identify the tail instance)
+    /// and the response itself.
+    pub tail_response_buffer: Rc<RefCell<Vec<(GlobalId, TailResponse)>>>,
+    /// Frontier of sink writes (all subsequent writes will be at times at or
+    /// equal to this frontier)
+    pub sink_write_frontiers: HashMap<GlobalId, Rc<RefCell<Antichain<Timestamp>>>>,
+    /// Peek commands that are awaiting fulfillment.
+    pub(crate) pending_peeks: Vec<PendingPeek>,
+    /// Tracks the frontier information that has been sent over `response_tx`.
+    pub reported_frontiers: HashMap<GlobalId, Antichain<Timestamp>>,
+    /// Undocumented
+    pub sink_metrics: SinkBaseMetrics,
+    /// The logger, from Timely's logging framework, if logs are enabled.
+    pub materialized_logger: Option<logging::materialized::Logger>,
+}
+
+/// A wrapper around [ComputeState] with a live timely worker and response channel.
+pub struct ActiveComputeState<'a, A: Allocate> {
+    /// The underlying Timely worker.
+    pub timely_worker: &'a mut TimelyWorker<A>,
+    /// The compute state itself.
+    pub compute_state: &'a mut ComputeState,
+    /// The channel over which frontier information is reported.
+    pub response_tx: &'a mut mpsc::UnboundedSender<Response>,
+}
+
+impl<'a, A: Allocate> ActiveComputeState<'a, A> {
+    /// Entrypoint for applying a compute command.
+    pub(crate) fn handle_compute_command(&mut self, cmd: ComputeCommand) {
+        match cmd {
+            ComputeCommand::CreateInstance | ComputeCommand::DropInstance => {
+                // Can be ignored for the moment.
+                // Should eventually be filtered outside of this method,
+                // as we are already deep in a specific instance.
+            }
+            ComputeCommand::CreateDataflows(_dataflows) => {
+                unreachable!("CreateDataflows should not be issued directly to compute state")
+            }
```

Comment:
The `CreateInstance`, `DropInstance`, and `CreateDataflows` commands seem to be unicorns among the other commands as they describe orthogonal concepts. We could ponder on extracting a different `Command` variant to capture that and avoid the `unreachable!` match arms. What do you think? This is not meant to be criticism of the current PR but follow-up work.

Reply:
Ah, I'm not sure yet. They are each new enough that idk what they end up looking like. Probably the create and drop instance commands don't survive here, but .. create dataflows would probably be fine except that we need to pick out the source creation from the dataflow, a thing that could/should actually happen with `StorageCommand::CreateSource`, or by compute pinging storage on construction.